src/Groups/abelian_aut.jl

```diff
@@ -160,3 +159,117 @@ If `M` defines an endomorphism of `G`, return `true` if `M` defines an automorph
+#
+#   Special functions for orthogonal groups of torsion quadratic modules
+#
+################################################################################
+
+
+"""
+    orthogonal_group(T::TorQuadMod, gensOT::Vector{fmpz_mat}) -> AutomorphismGroup{TorQuadMod}
+
+Return the subgroup of the orthogonal group of `G` generated by `gensOT`.
+"""
+function orthogonal_group(T::TorQuadMod, gensOT::Vector{fmpz_mat}; check=true)
+  Ggap, to_gap, to_oscar = _isomorphic_gap_group(abelian_group(T))
+  A = abelian_group(T)
+  function toA(x)
+    return A(x)
+  end
+  function toT(x)
+    return T(x)
+  end
+  T_to_A = Hecke.map_from_func(toA, T, A )
+  A_to_T = Hecke.map_from_func(toT, A, T)
+  to_oscar = compose(to_oscar, A_to_T)
+  to_gap = compose(T_to_A, to_gap)
+  AutGAP = GAP.Globals.AutomorphismGroup(Ggap.X)
+  ambient = AutomorphismGroup{typeof(T)}(AutGAP, T)
+  set_attribute!(ambient,:to_gap => to_gap)
+  set_attribute!(ambient,:to_oscar => to_oscar)
+  gens_aut = GapObj([ambient(g, check=check).X for g in gensOT])  # performs the checks
+  if check
+    # expensive for large groups
+    subgrp_gap =GAP.Globals.Subgroup(ambient.X, gens_aut)
+  else
+    subgrp_gap =GAP.Globals.SubgroupNC(ambient.X, gens_aut)
+  end
+  aut = AutomorphismGroup{typeof(T)}(subgrp_gap, T)
+  set_attribute!(aut,:to_gap => to_gap)
+  set_attribute!(aut,:to_oscar => to_oscar)
+  return aut
+end
+
+function Base.show(io::IO, aut::AutomorphismGroup{TorQuadMod})
+  T = domain(aut)
+  print(IOContext(io, :compact => true), "Group of isometries of ", T , " generated by ", length(gens(aut)), " elements")
+end
+
+
+"""
+    matrix(f::AutomorphismGroupElem{TorQuadMod}) -> fmpz_mat
+
+Return a matrix inducing `f`.
+"""
+matrix(f::AutomorphismGroupElem{TorQuadMod}) = hom(f).map_ab.map
+
+"""
+    defines_automorphism(G::TorQuadMod, M::fmpz_mat) -> Bool
+
+If `M` defines an endomorphism of `G`, return `true` if `M` defines an automorphism of `G`, else `false`.
+"""
+function defines_automorphism(G::TorQuadMod, M::fmpz_mat)
+  g = hom(G, G, M)
+  if !isbijective(g)
+    return false
+  end
+  # check that the form is preserved
+  B = gens(G)
+  n = length(B)
+  for i in 1:n
+    if Hecke.quadratic_product(B[i]) != Hecke.quadratic_product(g(B[i]))
+      return false
+    end
+    for j in 1:i-1
+      if B[i]*B[j] != g(B[i])*g(B[j])
+        return false
+      end
+    end
+  end
+  return true
+end
+
+function Base.show(io::IO, f::AutomorphismGroupElem{T}) where T<:TorQuadMod
+  if get(io, :compact, false)
+    print(io, matrix(f)) #somehow does not work
+  else
+    D = domain(parent(f))
+    print(io, "isometry ", D, " -> ", s,": ", matrix(f))
+  end
+end
+
+"""
+    orthogonal_group(T::TorQuadMod)  -> AutomorphismGroup{TorQuadMod}
+
+Return the full orthogonal group of this torsion quadratic module.
+
+May fail if T is too degenerate.
+"""
+function orthogonal_group(T::TorQuadMod)
+  OT = get_attribute(T, :orthogonal_group)
```

Comment:
This causes an error in CI because `TorQuadMod` does not store attributes. Workaround: add `@attributes TorQuadMod` somewhere (and then submit a patch to add it to the type definition in Hecke, if you haven't already; once that is released and we depend on that Hecke version, we can remove the `@attributes TorQuadMod` here again)